### PR TITLE
Allow User.hashPassword/validatePassword to be overridden

### PR DESCRIPTION
**Pull request by raymondfeng**
/to @ritch @bajtos 

The PR allows User.hashPassword and validatePassword methods to be overridden. It can be used as a workaround to https://github.com/strongloop/loopback/issues/251.

### CODE REVIEW

**Comment by bajtos** [comment]
What is the scenario where users should override `User.hashPassword`? I am concerned that most developers don't understand security implications of different hashing algorithms and thus can easily introduce a serious vulnerability.

File: common/models/user.js [hunk 1]
```diff
@@ -454,6 +454,22 @@ module.exports = function(User) {
...
+    this.validatePassword(plain);
+    var salt = bcrypt.genSaltSync(this.settings.saltWorkFactor || SALT_WORK_FACTOR);
+    return bcrypt.hashSync(plain, salt);
+  };
+
+  User.validatePassword = function(plain) {
+    if (typeof plain === 'string' && plain) {
+      return true;
+    }
+    throw new Error('Invalid password: ' + plain);
```
Comment (bajtos) [inline common/models/user.js]: How `throw new Error` going to integrate with the existing validations? The request setting invalid password should return HTTP status 422 and a description of the validation error in the usual format. This code will IMO return 500 without enough details.

**Comment by bajtos** [comment]
LGTM.

**Comment by bajtos** [comment]
please check why the CI builds are failing before you land this patch.

**Comment by cleas** [comment]
User.create({...}, function (err,user){...} ) will cause an uncaught exception as  
validatePassword throws an error if the supplied password is invalid. This is a change in behavior , I would expect the error to be returned in the callback function.

**Comment by raymondfeng** [comment]
Using the setter to hash passwords is pretty evil. We should find a better solution. Maybe the validatePassword should be set up as a validator on the User model.

**Comment by bajtos** [comment]
Once we have "persist" hook in place (https://github.com/strongloop/loopback-datasource-juggler/issues/559), it should be very easy to rework the User model to hash the password in this hook.

Let's keep the scope of this discussion around the change in behaviour introduced by this pull request. @cleas can you post an example that works in `loopback@2.8` and fails in `loopback@2.9`?

**Comment by cleas** [comment]
example : https://github.com/cleas/loopback-sandbox

**Comment by bajtos** [comment]
@cleas can you please open a new github issue to track this problem? I'll post some ideas how to fix it there.

**Pull request merged**